fix ipcalc.sh issues

yogo1212 opened the pull request:
Fixes: https://github.com/openwrt/openwrt/issues/12921
https://forum.openwrt.org/t/openwrt-23-05-0-rc1-first-release-candidate/162544/158?u=yogo1212

Still missing:
eval return code from subshell

yogo1212 committed:
base-files: ipcalc.sh: correctly interpret num parameter

The start and end addresses are inclusive.
Thus, adding num without substracting one results in num + 1 addresses.
Add the substraction and to implement the documented behaviour.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 committed:
base-files: ipcalc.sh: correctly set limit for prefix > 30

For /31 and /32 prefixes, there are only host addresses - no network and
broadcast address with all-zero and all-one bits.
Reflect this when setting the limit.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 committed:
base-files: ipcalc.sh: don't print broadcast addr for prefix > 30

Printing a broadcast address doesn't make any sense for /31 and /32
prefixes.
Strictly speaking, the same goes for the network address but it is useful
to get the first address in the prefix, e.g. to create a canonical
CIDR notation "$NETWORK/$PREFIX".

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 committed:
package: dnsmasq: remove off-by-one mitigation for limit

In the dnsmasq init script, an off-by-one in the range calculation of
ipcalc.sh was mitigated by passing the limit as if its counting started
at zero. This patch removes the mitigation as the off-by-one has been
fixed.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 commented:
Added a crude way to pass the success state of ipcalc.sh through eval.
Removed `BROADCAST` from the output for prefixes >30.

dave14305 commented:
Some additional observations while looking into this:

- There is no feedback/logging to the user to know they have an issue with their configuration. Maybe we need to send the stderr output from ipcalc.sh to syslog?
- The `START` variable is also used at the top of the init script to indicate startup order (19). This $START value was leaking into the dhcp-range specification when ipcalc failed. It shouldn’t be an issue if the return code is respected, but maybe local START, END, NETMASK variables should be declared in `dhcp_add()`.

Ansuel commented:
@yogo1212 is this something that fix one of the commit here?

https://github.com/openwrt/openwrt/pull/9857

john-tho commented:
Thank you for digging into this so fast!

> @yogo1212 is this something that fix one of the commit here? #9857

I think the main fix here is Fixes: e4bd3de1be8e ("dnsmasq: refuse to add empty DHCP range"). The commit message has the logic we want, but the code does not achieve it:
```
~ $ eval "$(false)"
~ $ echo $?
0
~ $ eval "false"
~ $ echo $?
1
```

---

The other ipcalc.sh users I can find only use NETWORK:
- https://github.com/openwrt/openwrt/blob/c4be106f4d989beae7e6dc0066c0595e1b1caa00/package/network/config/netifd/files/lib/netifd/dhcp.script#L22
- https://github.com/openwrt/openwrt/blob/c4be106f4d989beae7e6dc0066c0595e1b1caa00/package/network/config/netifd/files/lib/netifd/dhcp.script#L27
- https://github.com/openwrt/openwrt/blob/c4be106f4d989beae7e6dc0066c0595e1b1caa00/package/network/ipv6/6rd/files/6rd.sh#L44

One package uses NETMASK & IP:
- https://github.com/openwrt/packages/blob/73ee68a5970d1618cdb580001969d13a82e1e76d/net/isc-dhcp/files/dhcpd.init#L413

I think these uses were okay, because those vars were still printed and evaluated before ipcalc.sh could exit (unless args wrong).

Cheers

john-tho commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -50,12 +50,14 @@ BEGIN {
 	print "NETWORK="int2ip(network)
+	if (prefix<31) {
+		broadcast=or(network,compl32(netmask))
+		print "BROADCAST="int2ip(broadcast)
```
The one (packages feed: https://github.com/openwrt/packages/blob/73ee68a5970d1618cdb580001969d13a82e1e76d/net/isc-dhcp/files/dhcpd.init#L353) user I found declares `local BROADCAST` within it's `config_foreach` function, so no issue with using `BROADCAST` of a previous iteration, but it may need adjusting so that it also tests `[ -n $BROADCAST]` if we do not print it.

yogo1212 replied:
Good find!
I missed that.

This is the PR to address it:
https://github.com/openwrt/packages/pull/21404

yogo1212 committed:
package: avoid the use of eval to parse ipcalc.sh output

Add a function 'ipcalc' to /lib/functions.sh that sets variables more
safely using export.
With this new function, dnsmasq also handles the return value of ipcalc
correctly.

Fixes: e4bd3de1be8e ("dnsmasq: refuse to add empty DHCP range")
Co-Authored-By: Philip Prindeville <philipp@redfish-solutions.com>
Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 commented:
@dave14305 

> The START variable is also used at the top of the init script to indicate startup order (19). This $START value was leaking into the dhcp-range specification when ipcalc failed. It shouldn’t be an issue if the return code is respected, but maybe local START, END, NETMASK variables should be declared in dhcp_add().

The variable re-use/shadowing can only be properly solved in a subshell.
Sadly, the variables are used later in some other functions as well (-> can't use subshell and `local` - [depending on the shell](https://www.austingroupbugs.net/bug_view_page.php?bug_id=767)).
This is just what life is like when using eval :-/

```
ipcalc.sh .. | (
  eval "$(cat)"

  # xappend exists but only changes state in the subshell
  xappend "--dhcp-range=$tags$nettag$START,$END,$NETMASK,$leasetime${options:+ $options}"
)

# or

ipcalc_shell .. <<EOF
[ "$dynamicdhcpv4" = "0" ] && END="static"

# if ipcalc_shell is function, same as above. if it's a script: xappend isn't available.
xappend "--dhcp-range=$tags$nettag\$START,\$END,\$NETMASK,$leasetime${options:+ $options}"
EOF
```

Other than that, I can't think of any solutions. Everything is going to be a compromise.
The compromise of the status quo (eval in the current shell) is that `ipcalc.sh` can mess with variables to save callers from having to name all variables themselves.

Another sort-of issue would be that the variables are upper case, which indicates they are environment variables.
But changing that is costly, so we would have to agree that's its worth the effort before I do anything with it.

How about adding an `IPCALC_PREFIX=LEASE_` variable to control the names of the 'output variables'?
It's more of a clutch - rather than a solution but it would give users a way to avoid naming conflicts.

> There is no feedback/logging to the user to know they have an issue with their configuration. Maybe we need to send the stderr output from ipcalc.sh to syslog?

If I'm not mistaken, netifd/procd is already sending stderr to syslog - I might be wrong, though.

@john-tho

Thanks for the feedback and looking for the commit :-)

@Ansuel 

I'm adding the `Fixes:` line.

There's a new commit that ditches `eval` where it's not needed.

yogo1212 commented:
Waiting for https://github.com/openwrt/packages/pull/21404 to be merged

yogo1212 commented:
bumped package versions here as well

john-tho commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -509,6 +509,12 @@ dhcp_boot_add() {
+	local output
+	if output="$(ipcalc.sh "$@")"; then
+		eval "$output"
+	fi
```
Still need to forward the return code of ipcalc somehow. Possibly
```suggestion
	else
		return 1
	fi
```

Should maybe also use this for packages/isc-dhcp (or another way of testing ipcalc.sh return), because it does use `$START` and `$END` (I missed this first look, sorry) https://github.com/openwrt/packages/blob/e257404154a78a45bba3919a3b1bf3c2b9b1aa72/net/isc-dhcp/files/dhcpd.init#L413

john-tho commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -509,6 +509,10 @@ dhcp_boot_add() {
 }
 
+eval_ipcalc() {
+	local output="$(ipcalc.sh "$@")" || return
```
I needed the `local output` on a separate line to the `output=` for this to function as expected.

john-tho commented:
Tested, and fixes dnsmasq crash loop. Thanks!
I picked current PR (https://github.com/openwrt/openwrt/commit/dbd5f272a3b340b3b32a7c4814a77314b909017e) into 23.05, then copied `init.d/dnsmasq` and `ipcalc.sh` onto my 23.05rc router.

---
Possibly found more issues (**out of scope of this PR**), just want to reference them:

`dnsmasq.init` subtracts 1 from limit: https://github.com/openwrt/openwrt/blob/acd9981b4ef750544202df9d9e2d0143a6dfd478/package/network/services/dnsmasq/files/dnsmasq.init#L586 so I need to use `limit 2` in dhcp config, otherwise dnsmasq restart gives:
```
ipcalc.sh 10.53.11.0 31 1 0
network (10.53.11.0/31) too small
```
but `dnsmasq` runs fine, with the `dhcp-range` missing from dnsmasq config, as expected.

With `start 1` `limit 2`, dnsmasq successfully provides a /31 address to a client, from config stanza:
```
dhcp-range=set:trusted,10.53.11.1,10.53.11.1,255.255.255.254,12h
dhcp-option=trusted,option:dns-server,0.0.0.0
```

Before the ipcalc.sh exit error changes, dnsmasq did work with `start 1` `limit 1`, with the start and end addresses reversed in the dnsmasq config stanza (guessing maybe dnsmasq detected host using the `.0` address, so assigned client the `.1`):
`dhcp-range=set:test,10.53.100.1,10.53.100.0,255.255.255.254,1h`

---
OpenWrt interface uses full broadcast `brd 255.255.255.255`:
```sh
root@shed:~# ip -4 a show dev br0.11
32: br0.11@br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    link/ether c4:ad:34:e6:25:b1 brd ff:ff:ff:ff:ff:ff
    inet 10.53.11.0/31 brd 255.255.255.255 scope global br0.11
       valid_lft forever preferred_lft forever
```
vs manually added
```sh
root@shed:~# ip a d 10.53.11.0/31 dev br0.11
root@shed:~# ip a a 10.53.11.0/31 dev br0.11
root@shed:~# ip -4 a show dev br0.11
32: br0.11@br0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
    inet 10.53.11.0/31 scope global br0.11
       valid_lft forever preferred_lft forever

```
I will look into `netifd` for this

---
`dnsmasq-dhcp` uses OpenWrt hostname if an existing host has device name via static dhcp assignment, example: `john.example.com` in `/etc/config/dhcp`:
```
root@shed:~#  logread -e dnsmasq | grep 10.53
Sun Jun 25 06:05:58 2023 daemon.warn dnsmasq-dhcp[1]: not giving name john.example.com to the DHCP lease of 10.53.11.100 because the name exists in /tmp/hosts/dhcp.cfg01411c with address 10.53.11.114
cat /var/dhcp.leases | grep shed
1687680376 74:aa:aa:aa:aa:41 10.53.11.1 shed
```

yogo1212 committed:
base-files: ipcalc.sh clarify range parameters in comment

To avoid confusion when working with ipcalc.sh, clarify that the last two
parameters belong to the range calculation and rename 'num' to the slightly
less ambiguous 'size'.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

yogo1212 commented:
@john-tho Thank you for the elaborate feedback!

It looks to me as if the calculation/usage of `limit` is still an open point that needs to be addressed:
Effectively, `limit` is documented to be the "size of the address pool" - a "count".
`ipcalc.sh` only has this comment to document its behaviour:

```
ipcalc <ip> <netmask> <start> <num>
```

To me, that corresponds to the `limit` option from the [UCI wiki page](https://openwrt.org/docs/guide-user/base-system/dhcp).
So, the off-by-one was fixed in `ipcalc.sh` and the mitigation in dnsmasq needs to go.

`isc-dhcp` didn't do `-1` either.

I've added two commits which I hope improve the situation.

Ansuel commented:
@yogo1212 so what is the situation with this? also related to the package in feeds?

yogo1212 commented:
rebased, fixed typo in commit message, and pinged @pprindeville on isc-dhcp

yogo1212 commented:
[In the forums](https://forum.openwrt.org/t/dhcp-doesnt-work-anymore-when-router-ip-is-inside-dhcp-range/166539/49), there's a user who's configuration stopped working because there is now a check that rejects configurations where the server address lies withing the DHCP range.
That worked correctly because dnsmasq skips addresses that the server is using.

I would like to add a fix for that functionality by adding a switch (environment) to output a list of ranges and split the range at the server address. Let me know if that should go into a separate PR. For now, I'm putting the commit here.

yogo1212 commented:
```
£ RANGES=1 package/base-files/files/bin/ipcalc.sh 192.168.12.12/25 2 125
IP=192.168.12.12
NETMASK=255.255.255.128
NETWORK=192.168.12.0
BROADCAST=192.168.12.127
PREFIX=25
RANGES=192.168.12.2,192.168.12.11;192.168.12.13,192.168.12.126
```

Ansuel commented:
Np for keeping stuff here but we need to understand what to backport to fix
the regression

Il Mar 25 Lug 2023, 12:31 Leon Busch-George ***@***.***> ha
scritto:

> £ RANGES=1 package/base-files/files/bin/ipcalc.sh 192.168.12.12/25 2 125
> IP=192.168.12.12
> NETMASK=255.255.255.128
> NETWORK=192.168.12.0
> BROADCAST=192.168.12.127
> PREFIX=25
> RANGES=192.168.12.2,192.168.12.11;192.168.12.13,192.168.12.126
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/openwrt/pull/12925#issuecomment-1649568762>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AE2ZMQTMBPFFDPAKZCAHCFDXR6OBTANCNFSM6AAAAAAZJHJ4A4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

yogo1212 commented:
Sure thing!

The user had dnsmasq configured for a network with the primary IP address `192.168.13.13/25` and start/limit set to `1/160`.
Previously, ipcalc didn't have the check prevent such configurations where the server's IP address lies within the DHCP range.
With dnsmasq, that wasn't an issue because dnsmasq checks its own address before handing out a lease.

ipcalc.sh is not only used for dnsmasq and overlapping the DHCP range with  is a bad practice. Afaics, isc-dhcpd doesn't have such a check.
The regression in the release should be fixed by backporting [base-files: ipcalc.sh: add option to emit multiple ranges](https://github.com/openwrt/openwrt/pull/12925/commits/80ed7caf542f8b0d9dafe1778de390ebdba68e92) and [package: dnsmasq: parse multiple ranges from ipcalc.sh](https://github.com/openwrt/openwrt/pull/12925/commits/73ac2b8269946fe985b1b7e0874f1b2349730942) from this PR.

yogo1212 commented:
Let me try to open a PR into the release branch.

dave14305 commented:
This just seems to be getting more complicated.

yogo1212 commented:
@dave14305 it is :-(

I would say that it's mostly down to the nature of the rudimentary ex-shell script, the output of which was passed to eval.
If it hadn't been so laissez-faire, we wouldn't have to think about supporting all sorts of crazy configurations.
Who uses a DHCP server in a /30 network? Who puts their server address inside the DHCP range it serves?

Remember that my patches from last year were also addressing issues like these (and dealing with faulty configurations that shouldn't brake the router's name resolution).
At this point, we have to go by the "don't break userspace" mantra.

If you have a good idea to make this thing easier to work with, I'm willing to put in some time.

dave14305 commented:
What dnsmasq problem was the original commit for `ipaddr inside range` trying to solve, since dnsmasq already handles the server address being in the range?

The script went 9 years without any updates, so I would think you’d want to tread lightly with it.

zdz2019 commented:
After testing, dhcpv4 is broken, and all clients cannot obtain dhcpv4 addresses (including wireless clients).

yogo1212 commented:
@zdz2019 can you post the relevant parts of your config and/or error information?
were you testing the release branch or this one?

zdz2019 commented:
Used the master branch, after replacing ipcalc.sh and dnsmasq.init it sends this error.

Sat Jul 29 18:10:23 2023 daemon.info hostapd: phy1-ap0: STA 00:00:00:00:00:00 IEEE 802.11: authenticated
Sat Jul 29 18:10:23 2023 daemon.info hostapd: phy1-ap0: STA 00:00:00:00:00:00 IEEE 802.11: associated (aid 1)
Sat Jul 29 18:10:23 2023 daemon.notice hostapd: phy1-ap0: AP-STA-CONNECTED 00:00:00:00:00:00 auth_alg=open
Sat Jul 29 18:10:23 2023 daemon.info hostapd: phy1-ap0: STA 00:00:00:00:00:00 RADIUS: starting accounting session 0CD7293AC364AE57
Sat Jul 29 18:10:23 2023 daemon.info hostapd: phy1-ap0: STA 00:00:00:00:00:00 WPA: pairwise key handshake completed (RSN)
Sat Jul 29 18:10:23 2023 daemon.notice hostapd: phy1-ap0: EAPOL-4WAY-HS-COMPLETED 00:00:00:00:00:00
Sat Jul 29 18:10:24 2023 daemon.warn odhcpd[2301]: No default route present, overriding ra_lifetime!

yogo1212 commented:
Your log output doesn't contain any related errors, imho.
The line `No default route present, overriding ra_lifetime!` has to do with IPv6 (ipcalc.sh only deals with IPv4).
To my knowledge, odhcpd isn't related to ipcalc.sh at all.
Are you using dnsmasq to serve DHCP?

Can you post your `dhcp` and `network` configs from /etc/config/?

Here's an excerpt of a config using dnsmasq:

```
# network
config interface 'test'    
        option force_link '1'
        option device 'br-test'
        option proto 'static'   
        list ipaddr '100.65.0.1/24'

# dhcp
config odhcpd 'odhcpd'
	option maindhcp '0'

config dhcp 'test'
        option interface 'test'
        option force '1'      
        option start '20'
        option limit '180'
        option leasetime '30m'
```

Results in this dnsmasq config:

```
dhcp-range=set:test,100.65.0.20,100.65.0.199,255.255.255.0,30m
```

There are no errors on my end. I can connect to an associated SSID with my phone and use the internet.

zdz2019 commented:
Simply replace ipcalc.sh and dnsmasq.init to reproduce the problem, no post-compile testing required.

You can verify the result after replacing it yourself.
/bin/ipcalc.sh
/etc/init.d/dnsmasq

/etc/config/ file is irrelevant to solve the problem, since you need to see it, then see below:

**/etc/config/dhcp**
config dnsmasq
	option enabled '1'
	option domainneeded '1'
	option localise_queries '1'
	option local '/home.arpa/'
	option domain 'home.arpa'
	option expandhosts '1'
	option authoritative '1'
	option readethers '1'
	option leasefile '/tmp/dhcp.leases'
	option resolvfile '/tmp/resolv.conf.d/resolv.conf.auto'
	option localservice '1'
	option rebind_protection '1'
	option rebind_localhost '1'
	option ednspacket_max '1232'
	option dnsforwardmax '1000'
	option cachesize '10000'
	list notinterface 'pppoe-wan'
	option bind 'dynamic'

config dhcp 'lan'
	option interface 'lan'
	option start '1'
	option limit '254'
	option leasetime '1d'
	option dhcpv6 'server'
	option ra 'server'
	option ra_management '1'

config dhcp 'wan'
	option interface 'wan'
	option ignore '1'

config odhcpd 'odhcpd'
	option maindhcp '0'
	option leasefile '/tmp/hosts/odhcpd'
	option leasetrigger '/usr/sbin/odhcpd-update'
	option loglevel '4'


**/etc/config/network**
config interface 'loopback'
	option device 'lo'
	option proto 'static'
	option ipaddr '127.0.0.1'
	option netmask '255.0.0.0'

config globals 'globals'
	option ula_prefix 'fd07:9f81:8a82::/48'
	option packet_steering '1'

config device
	option name 'br-lan'
	option type 'bridge'
	list ports 'lan1'
	list ports 'lan2'
	list ports 'lan3'
	list ports 'lan4'

config interface 'lan'
	option device 'br-lan'
	option proto 'static'
	option ipaddr '192.168.1.1'
	option netmask '255.255.255.0'
	option ip6assign '60'

config interface 'wan'
	option device 'wan'
	option proto 'dhcp'

config interface 'wan6'
	option device 'wan'
	option proto 'dhcpv6'

pprindeville commented:
@yogo1212 By the way, have you taken before/after synthesized configs and `diff`'d them to make sure there's no change?

yogo1212 commented:
Please excuse the trouble and my reaction, I'm having a hard time in my relationship right now.

I had `BUSYBOX_CONFIG_ASH_BASH_COMPAT` enabled.
Without it, the `<()` doesn't work.
Previously, I had a pipe going into a `while` loop which I abandoned because it runs in a subshell. I'll switch to the pipe and hopefully that fixes it.

yogo1212 commented:
I'm surprised there was no error concerning the shell syntax in the syslog.

@pprindeville yes, the md5sums of /tmp/etc/dnsmasq* match with and without the patches, except for those cases where the range was split or the calculations were faulty before.

mpratt14 commented:
> I had a pipe going into a `while` loop

perhaps a cleaner way would be a for loop and string subsitutions

```
for range in ${RANGES//;/ }; do
	range_start=${range%%,*}
	range=${range#*,}
	range_end=${range%%,*}
	...
```

I'd also like to introduce a shell function for word spltting... unless we have one already

yogo1212 commented:
@mpratt14 Very good idea!
I was fiddling around with IFS and such but it's difficult in that construct.

mpratt14 commented:
are you certain that there are only two comma-separated items in each `range`? otherwise you would have to redefine it in order to get the 2nd item instead of the last item. I don't know how we do it here I just wrote my example based on dnsmasq docs which has the leasetime after end of range, and with several optional items in between

Also, it would be nice to find a way to not have to add usage of sed

yogo1212 commented:
@mpratt14 The `range` variable is derived from `RANGES` which comes from `ipcalc.sh`. It is independent of dnsmasq. So for our purposes here, ranges would be defined only by their first and last addresses.


Concerning `sed`:
There were worries regarding the use of eval in lines like
`eval "$(ipcalc.sh "$ip/$mask")";ip_net="$NETWORK"`.
This was my proposal to do the same without eval:
`ip_net="$(ipcalc.sh "$ip/$mask" | sed -nE 's/^NETWORK=//p')"`
Maybe something like this is a reasonable compromise:
`ip_net="$(eval "$(ipcalc.sh "$ip/$mask")"; echo "$NETWORK")"`

mpratt14 commented:
there should be a way to avoid both sed and eval, but I'm not sure what's going on there since I'm not sure how ipcalc works. so what is ipcalc outputting there exactly?

yogo1212 commented:
```
$ ipcalc.sh "192.168.0.2/24" 5 210
IP=192.168.0.2
NETMASK=255.255.255.0
NETWORK=192.168.0.0
BROADCAST=192.168.0.255
PREFIX=24
START=192.168.0.5
END=192.168.0.214
```

Likely, ipcalc.sh was built with eval in mind. I, personally, don't mind eval that much here but it was brought up during further up in this PR. Then again, I don't mind sed either.
`START` is also used by rc.d. By the time ipcalc.sh is called, that is irrelevant though.
Other options to deal with multiple outputs could be json (jshn.sh also uses eval a lot but with prefixes to avoid naming conflicts). It would also be possible to store files in a temporary dir. Or to output only one parameter per call (i don't think ipcalc.sh is on any critical code path).

mpratt14 commented:
> `START` is also used by rc.d.

so it's meant to be global variables, right?

what if you just take each line and `export`

yogo1212 commented:
There appear to have been some shell calculations (e1bb37f2ec7aea568eee32a0c88f937a2b963c3f). Also, the file was called `ipcalc` at some point but rename to fix a bug (`rename /bin/ipcalc to /bin/ipcalc.sh (fixes #1452)`). What that bug was eludes me because bugs.openwrt.org is now a redirect to github and the web archive doesn't have https://bugs.openwrt.org/index.php?do=details&task_id=1452. Other components appear to have been using awk as well (e.g. 60c1f0f64d23003a19a07d6b9638542130f6641d mentions awk with uci). 

If a large enough amount of people would prefer to get rid off awk and be able to change the interface, we could go for something new. However, ipcalc.sh probably needs to stay to not break dependencies.

I got to say, though, that I only opened this PR to fix some bugs.
Imho, the use `eval` is undesirable here for one main reason: It's hard to know which pre-existing variables it might overwrite.
`sed` doesn't bother me at all - it's  `grep .. | cut -d= ..` in one and avoids nesting or extra lines.

The use of upper case variable names in the output suggests 'environment' but I don't think environment variables are a good choice given the way how ipcalc is used (very local - I want to limit the scope, not broaden it).
Requiring the caller to iterate all output lines and call `export` or `set` doesn't seem appropriate to me.

Imo, the biggest issue is the lack of control over or transparency of the variable names that ipcalc.sh outputs.

I don't think that is possible without a new interface.
How about a shell function 'ipcalc_set' that sets select local variables based on a prefix?

dnsmasq could call it like this:
```shell
. /lib/functions.sh
..
local dhcp_netmask dhcp_ranges
..
PREFIX='dhcp' USE_RANGES=1 ipcalc_set netmask,ranges 192.168.1.4/24 1 254
# dhcp_netmask = 255.255.255.0
# dhcp_ranges = 192.168.1.1,192.168.1.3;192.168.168.1.5,192.168.1.254
```

```shell
local ip_net
ip_net="$(ipcalc.sh "$ip/$mask" | sed -nE 's/^NETWORK=//p')"
# would become
. /lib/functions.h
..
local ip_netmask
PREFIX='ip' ipcalc_set netmask "$ip/$mask"
```

This new interface can use ipcalc.sh internally, convert the output variable names to lower case, filter the output lines for those selected by the user, and set variables:

```shell
ipcalc_set() {
  local filter="${1//,/$IFS}" ; shift

  local fifo="$(mktemp -u "/tmp/ipcalc_$$.XXXXXX")"
  mkfifo "$fifo"

#   # create and immediately delete a fifo to avoid leaking the file
#   exec 6<>"$fifo"
#   rm "$fifo"

  ipcalc.sh "$@" >"$fifo" &
  local pid=$!

  local req val var
  while read var; do
    val="${var#*=}"
    var="$(echo "${var%%=*}" | awk '{print tolower($0)}')"

    for req in $filter; do
      [ "$req" = "$var" ] || continue

      eval "${PREFIX-ipcalc}_${var}=${val}"
      # alternatively, use read or ': ${:=}' to set
      # then, RANGES mustn't be quoted by ipcalc.sh
      break
    done
  done <"$fifo"

  rm "$fifo"

  wait "$pid"
}
```

Or ipcalc.sh is implemented using a new interface `ipcalc` that has an explicit option to generate output safe for use with eval.

Again, I'm not sure that's in scope of this PR.
Determining the interface for a replacement will already be a point of contention.

mpratt14 commented:
you can find the old bug reports, they are just more difficult to find 

FS#1452 is #6421

mpratt14 commented:
I think another function just to wrap ipcalc.sh is overkill, I don't see a reason that ipcalc.sh has to remain the same output format, one simply has to adjust all the dependencies at the same time one rewrites ipcalc.sh

But of course no one is expecting you to do a full rewrite here, you've done enough. This is just future ideas. I don't think export is any more harmful than the way it already is with eval to be honest.

if the output of ipcalc.sh is truly only needed locally in this case, I would just save the entire output to a variable on it's own and then bring it back as a prefix to whatever command needs the values. If I remember right the shell will automatically change the newlines to spaces (and if not you can do another string substitution or an IFS= before saving the variable of variables)

pprindeville commented:
> We don't know who is relying on ipcalc.sh (e.g. private repos) and the interface shouldn't be changed (or ipcalc.sh removed) without a deprecation warning and at least one release cycle.

This can't be emphasized too much.

dave14305 commented:
> What that bug was eludes me because bugs.openwrt.org is now a redirect to github and the web archive doesn't have https://bugs.openwrt.org/index.php?do=details&task_id=1452.

> That bug contains log messages from 2018. The commit with the line in question is from 2007 - maybe that's yet another bug tracker?

https://dev.archive.openwrt.org/ticket/1452

https://dev.archive.openwrt.org/changeset/6595.html

dave14305 commented:
At a basic level, what about just fixing the original intention to not add the dhcp-range if ipcalc.sh returns an error?

```bash
if [ "$dhcpv4" != "disabled" ] && eval "$(ipcalc.sh "${subnet%%/*}" "$netmask" "$start" "$limit" || echo false)" ; then
```

pprindeville commented:
> At a basic level, what about just fixing the original intention to not add the dhcp-range if ipcalc.sh returns an error?
> 
> ```shell
> if [ "$dhcpv4" != "disabled" ] && eval "$(ipcalc.sh "${subnet%%/*}" "$netmask" "$start" "$limit" || echo false)" ; then
> ```

I'm inclined to agree that simpler is better, and that anything more structural should be made incrementally.

yogo1212 commented:
> > At a basic level, what about just fixing the original intention to not add the dhcp-range if ipcalc.sh returns an error?
> > ```shell
> > if [ "$dhcpv4" != "disabled" ] && eval "$(ipcalc.sh "${subnet%%/*}" "$netmask" "$start" "$limit" || echo false)" ; then
> > ```
> 
> I'm inclined to agree that simpler is better, and that anything more structural should be made incrementally.

I'm not attached to `eval_ipcalc` - we can remove it if you want.
I only added it because of concerns that the 'injection' (`echo false`) isn't reliable and depends heavily on the previous output of a process that is known to have failed (e.g. `NETMASK=255.255.255.0false`). The `awk` process can fail for reasons other than `exit(1)`.

yogo1212 commented:
> > What that bug was eludes me because bugs.openwrt.org is now a redirect to github and the web archive doesn't have https://bugs.openwrt.org/index.php?do=details&task_id=1452.
> 
> > That bug contains log messages from 2018. The commit with the line in question is from 2007 - maybe that's yet another bug tracker?
> 
> https://dev.archive.openwrt.org/ticket/1452
> 
> https://dev.archive.openwrt.org/changeset/6595.html

Nice! Thank you :-)
It doesn't contain the line of reasoning that I was hoping to find, though.
Still don't know why it had to become `.sh`. :shrug:

dave14305 commented:
Who is going to decide on the best course of action to fix 23.05 (hopefully before rc3 is tagged)? Push a fix or revert the “ipaddr in range” validation.

pprindeville commented:
> @dave14305 No idea who that would make that decision.
> 
> I have a feeling that the most contentious point is still how a configuration like `addr: 10.0.0.2/30, start: 1, limit: 3` should be viewed - something we should probably discuss before talking about what the code should look like.
> 
> To me, it's an addressing conflict that shouldn't be encouraged but has come to be expected to work by users. We're only talking about keeping or removing the check because dnsmasq has an undocumented (?) mitigation.
> 
> Dnsmasq will accept such a configuration and doesn't break the network. To my understanding, isc-dhcpd and odhcpd will accept the configuration and hand out their own addresses as leases (-> break the network). That should be accounted for and ipcalc is a good place to do that, imho. The check, by itself, makes sure thatl DHCP servers (generally) reject the configuration - if they use ipcalc. Odhcpd doesn't use ipcalc and would need to be patched separately.
> 
> [bc6f424](https://github.com/openwrt/openwrt/commit/bc6f424682d1f6aa5cc7c983e0503372915f7aea) re-enables the functionality for dnsmasq by adding an interface (ipcalc API, `USE_RANGES`) that supports emitting multiple ranges - other software (like isc-dhdpd) could make use of it as well.
> 
> If that is too complex, there could instead be two modes for ipcalc: Let's call them 'lax' (for dnsmasq) and 'strict' (for isc-dhcpd). With `LAX_RANGE=1`, ipcalc would skip `if (ipaddr > start && ipaddr < end)`, relying on dnsmasq built-in mechanism. I'm convinced that 'strict' is the correct default behaviour.
> 
> How does that sound to you?
> 
> Just reverting the change allows for faulty configurations to break the network.
> 
> I think that the avoidance of addressing conflicts is a task for the admin - aka the person configuring the network. Configuring the address locally AND as part of a DHCP address range is a duplicate allocation of an address. The script runs afterwards. There isn't going to be a reliable way of solving this kind of design flaw post-hoc. What if there are multiple local addresses? Dnsmasq only look at the first. If it were a new feature, I would clearly forbid it but since users expect the configuation to work.
> 
> Under the premise of leaving the responsibility with the admin, the check could result in a warning instead of an error. Only users who keep an eye on the log would see it, though. And handing out the gateway address can not only disturb the internet connection but could also prevent the admin from reaching the router to fix the configuration. So, I'm not in favour of this option and I don't think it aligns with the safety paradigms that Openwrt employs in other places. But, given the situation, it has some legitimacy. (Maybe make the check emit a deprecation warning now and turn it into an error after the next release?)

I don't mind the init scripts doing good sanity checking.  That's always helpful.  And indeed I'll try to bake a change into the next spin of DHCP that checks for the range spanning the server's own IP.  Didn't realize people could be so foolish, but...

I *do* object to baking knowledge of the quirks of `dnsmasq` into `ipcalc.sh`.  I don't see the point.  And what happens if `dnsmasq` ever changes its behavior?  Now we've got users of `ipcalc.sh` with expectations of continuity...

If `ipcalc.sh` is done properly, it should be usable from within the `dnsmasq` init script to make all the appropriate checks without having any additional complexity (or oddity, non-intuitive behavior, etc) baked into it.

yogo1212 committed:
base-files: ipcalc.sh: turn error into warning

Some users have their routers configured to supply a DHCP range that
includes the local interface address.
That worked with dnsmasq because it automatically skips the local
address.

Re-enable those existing configurations for the release and hint at
possible future problems.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>
[ wrap commit description and remove unecessary text ]
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

yogo1212 commented:
@pprindeville 

> I do object to baking knowledge of the quirks of dnsmasq into ipcalc.sh. I don't see the point. And what happens if dnsmasq ever changes its behavior? Now we've got users of ipcalc.sh with expectations of continuity...
>
> If ipcalc.sh is done properly, it should be usable from within the dnsmasq init script to make all the appropriate checks without having any additional complexity (or oddity, non-intuitive behavior, etc) baked into it.

My thoughts exactly! That's what the `USE_RANGES` env switch is supposed to achieve:
.. an interface that enables the functionality for all DHCP servers (not only dnsmasq).
All servers would add multiple ranges (-> two ranges, wrapping the local address).
Should dnsmasq ever drop their mitigation, the configuration generated using ipcalc should still be correct (by skipping the local address in the configuration rather than at runtime).

I take it that the complexity is still a contentious point. While I agree that it is undesirable, I don't see any good options that are less complex (at least none that maintain strict checks and use an interface that works for any server).

@dave14305 

> I’d be in favor of backporting “warning-only” mode to 23.05. Let main keep going with the hard error. My 2 cents.

Good point. I've updated the PR for the release branch to not `exit(1)` and removed the alterations to ipcalc and dnsmasq.init.
I'm not against going back to the hard error but then we need to instead think about making sure that all users with such configurations are warned and then wait for another release.

Coming to think about it, that would also be necessary for making ranges the default option and dropping the `START` and `END` output lines.

Ansuel commented:
@yogo1212 IMHO to handle these kind of wrong configuration the only way is to put some text in the release notes and also spam descriptive error code and maybe some link about it?

yogo1212 commented:
@Ansuel I've replaced the last two commits (concerning split ranges) with one that I hope meets your description.
It's unclear to me which resource could be linked to - my quick searches didn't bring up an appropriate resource.
I'm open to improve it, so please send suggestions.
Also, feel free to alter the commit for a merge to make the process easier :slightly_smiling_face:

pprindeville commented:
I don't have a problem with comments 1-3, or 6 & 7.  And indeed, we could have gotten those in quickly if the cleanup had been more narrowly scoped.

I don't like the use of `eval`, and other libraries/functions like `jshn.sh` use something like:
```
export -- "$varname1=\"$value1\"" \
          ...
```

yogo1212 commented:
@pprindeville 

That certainly looks a lot nicer!
At previous points in the discussion, we tried to avoid setting environment variables to limit the scope because names like `START`, `NETWORK`, and `END` are common words - as found with dnsmasq where the 'runlevel' variable `START` is overwritten.
For dnsmasq, I could imagine integrating your suggestion because it doesn't impact 'invisible' users.
Also, I just checked:
```
START=123
test() {
  local START
  ipcalc 192.168.0.1/24 2 24
}
echo "$START"
```
.. prints `123`, so it's possible to limit the scope using `local` even with env vars.

yogo1212 commented:
force pushing to resolve conflict (version increment netifd, i assume)

pprindeville reviewed (changes requested):
Still using `eval`... and other minor cleanup.

yogo1212 commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -50,12 +50,14 @@ BEGIN {
 	print "NETMASK="int2ip(netmask)
-	print "BROADCAST="int2ip(broadcast)
 	print "NETWORK="int2ip(network)
+	if (prefix<31) {
```
Minor grumble, maybe.. but easy to remedy

yogo1212 commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -89,10 +89,8 @@ BEGIN {
-		print "ipaddr inside range" > "/dev/stderr"
-		exit(1)
-	}
+	if (ipaddr > start && ipaddr < end)
```
The existing code already does `if (end==ipaddr) end=ipaddr-1` and `if (start==ipaddr) start=ipaddr+1`, though I agree the line smells funny.

Make a wish: change or keep?

yogo1212 replied:
Alright!

yogo1212 commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -509,6 +509,11 @@ dhcp_boot_add() {
 }
 
+eval_ipcalc() {
+	local output
```
The `eval_` was intented more as a warning about the internals :sweat_smile: 

I'm not entirely convinced of the use of environment variables.
Consider the use of `ipcalc.sh` for isc-dhcpd: If `BROADCAST` was set for a previous UCI section, we don't want that value to leak into subsequent calls of `dhcpd_add`.
For some reason I don't understand, `local v; export -- v=test` doesn't alter any environment variables outside of the function (at least with bash and busybox ash).

There [is a](https://github.com/openwrt/packages/pull/21404/files#diff-ac27427e7c4b64d6e9b62ac528b66be6c626d80ae38ddfad12c6a826a1c5fccdR412) `local .. BROADCAST` before the call to ipcalc. Either `poplocalvars` (seen [here](https://github.com/brgl/busybox/blob/098b713c7b5b22fc60b606be97da431eaedc6639/shell/ash.c#L9145)) restores environment variables or `setvareq` works in such a way that `export` can create an environment variable that is local to the function.

The 'lucky' behaviour of `local` is not documented anywhere that I could find.
It seems a little obscure to me and I prefer the `eval` because I imagine it's easier to understand what's going on.

pprindeville replied:
> If BROADCAST was set for a previous UCI section, we don't want that value to leak into subsequent calls of dhcpd_add.
For some reason I don't understand, local v; export -- v=test doesn't alter any environment variables outside of the function (at least with bash and busybox ash).

They’re not “environment” variables if they aren’t exported into another process; they’re just “variables”. Scoping them as `local` stops this from happening.

pprindeville replied:
There’s nothing “lucky” about how `local` works. It does exactly what it’s designed to. From `dash(1)`:

>  Functions
>      The syntax of a function definition is
> 
>            name () command
> 
>      A function definition is an executable statement; when executed it
>      installs a function named name and returns an exit status of zero.
>      The command is normally a list enclosed between “{” and “}”.
> 
>      Variables may be declared to be local to a function by using a
>      local command.  This should appear as the first statement of a
>      function, and the syntax is
> 
>            local [variable | -] ...
> 
>      Local is implemented as a builtin command.
> 
>      When a variable is made local, it inherits the initial value and
>      exported and readonly flags from the variable with the same name in
>      the surrounding scope, if there is one.  Otherwise, the variable is
>      initially unset.  The shell uses dynamic scoping, so that if you
>      make the variable x local to function f, which then calls function
>      g, references to the variable x made inside g will refer to the
>      variable x declared inside f, not to the global variable named 

x.

pprindeville replied:
> Consider the use of `ipcalc.sh` for isc-dhcpd: If `BROADCAST` was set for a previous UCI section, we don't want that value to leak into subsequent calls of `dhcpd_add`.

I’m the maintainer of `isc-dhcp` and I use stacked (nested) `local`’s without any issues. If need be, we can rename any variables that are problematic.

In any case, isc-dhcp is being retired as soon as I can finish getting the front end to synthesize Kea configuration instead.

pprindeville replied:
Lastly, `eval`’s each cost an extra fork/exec pair which is a *lot* slower than staying inside the same process. They’re to be avoided if there’s another way.

yogo1212 replied:
Ok. My biggest misunderstand was that `export` only marks variables for being exported when a command is run - rather than altering the environment. Insofar, `local` really doesn't do anything special in this case.

But I wasn't arguing the use of `eval`. I was pondering the use of `ip_net="$(ipcalc "$ip/$mask" && echo "$NETWORK")"`
instead of `ipcalc "$ip/$mask" && ip_net="$NETWORK"` because it requires only one `local ip_net` to avoid leaking or possibly overwriting foreign variables.

I hadn't yet considered fork/exec overhead because it's not on a critical code path, at least none that I know of.
Also, there was an `eval` there before, so we're not making anything worse.

pprindeville replied:
> I hadn't yet considered fork/exec overhead because this is not a critical code path, at least not to my knowledge.
> Also, there was an eval there before, so we're not making anything worse.

As it happens, the rewrite of the dhcp synthesizer is going to make heavy use of `ipcalc.sh` to figure out which subnet a host belongs in... otherwise you have to use a single global pool and the performance doesn't scale up well.

yogo1212 replied:
@pprindeville Ok, so there are plans for ipcalc. I can't tell from your description whether that makes it critical for performance (still only used during initialization?).

For 'heavy use', I recommend creating a new interface.
Shell is not well-suited for these 'dictionary' type return values. Best I can think of is handing an environment to a sub process (or to a function call, when using `source`). Ucode, on the other hand, has objects.
@ynezz talked about leveraging ucode's abilities and sharing code with fw4 to extend on [the other PR](https://github.com/openwrt/openwrt/pull/13158).

But is that in the scope of this PR?
Imho, this PR rather needs a decision on which one of these is more appropriate:

```shell
local ip_net
ip_net="$(ipcalc "$ip/$mask" && echo "$NETWORK")"
# or
local ip_net BROADCAST START END NETWORK NETMASK PREFIX
ipcalc "$ip/$mask" && ip_net="$NETWORK"
```

Also, is it worth adding the extra function in `dhcp.script` and `6rd.sh`?
Mt. St. Optimization solves the space elevator problem :stuck_out_tongue: 
I only wanted to fix a crash-loop a year ago.

I'm replacing `ipcalc_eval` with your neat `ipcalc` function for dnsmasq - low hanging fruit (https://github.com/openwrt/openwrt/pull/12925/commits/af098729edaba0d9e6a6a0363c1599fa898be662).
Then, we should only be talking about https://github.com/openwrt/openwrt/pull/12925/commits/7b6d9f98ef2653469f464ad01357c2df430c122f

pprindeville replied:
I wish busybox `awk` supported `-E` (I've opened a bug requesting it) so we could do what `ipcalc` on RedHat does (you can pass in flags to only get individual values... in this case, `-n` would limit the output to `NETWORK=`).

I'd go with the 2nd format.

Initialization doesn't always just happen at startup.  Re-initialization might be required if a watched interface bounces, or if the config file gets touched and needs to be reloaded, etc.

yogo1212 replied:
alright! force pushed.

yogo1212 commented on package/network/config/netifd/files/lib/netifd/dhcp.script:
```diff
@@ -19,12 +19,12 @@ setup_interface () {
 
 	local ip_net
-	eval "$(ipcalc.sh "$ip/$mask")";ip_net="$NETWORK"
+	ip_net="$(ipcalc.sh "$ip/$mask" | sed -nE 's/^NETWORK=//p')"
```
The reason I touched the code and introduced the subshell were comments the 'leaky' nature of eval.

```
ipcalc() {
    set -- $(ipcalc.sh "$@")
    [ $? -eq 0 ] && export -- "$@"
}
..
ipcalc "$ip/$mask" && ip_net="$NETWORK"
```
I like the added safety but tend to err more on the 'limit side-effects to subshell'-side of things.
How about `ip_net="$(ipcalc "$ip/$mask" && echo "$NETWORK")"`?

There we'd have a subshell to limit side-effects and no eval either

pprindeville replied:
There’s no risk of command injection if you just use the `export`

pprindeville replied:
> There we'd have a subshell to limit side-effects and no eval either

You’re misunderstanding how `$(…)` works. For a local shell function, it doesn’t create a subshell. It doesn’t need to.

But subshells have a *lot* of overhead and are to be avoided except when there’s no other way.

yogo1212 replied:
> There’s no risk of command injection if you just use the `export`

That, I understand. I was talking about not accidentally overwriting variables in that context which `$()` does help with, afaics (`: $(q=1234) ; echo "$q"`).

> For a local shell function, it doesn’t create a subshell.

Fascinating!

pprindeville replied:
Right. It doesn’t create a subshell because (1) it’s not necessary, (2) it would be extremely limiting in what you could do, and (3) it would be horribly inefficient.

Your example of an assignment inside of `$(…)` is a bit tortured. Why would anyone do that instead of just assigning without it?

If a function is well behaved and declares all of its variables as `local` then there’s no danger of overwriting variables in parents’ scope. So I’m unclear what the concern or hesitation is.

yogo1212 replied:
i've closed the 'dittos'.
please resolve this if you're happy with the change :-)

yogo1212 committed:
base-files: ipcalc.sh: make check slightly more future-proof

The previous code handling the equal-condition might be removed or
altered in the future and the case might be overlooked.

Signed-off-by: Leon M. Busch-George <leon@georgemail.eu>

pprindeville commented on package/base-files/files/bin/ipcalc.sh:
```diff
@@ -90,8 +90,7 @@ BEGIN {
 	if (ipaddr > start && ipaddr < end) {
-		print "ipaddr inside range" > "/dev/stderr"
-		exit(1)
+		print "warning: ipaddr inside range - this might not be supported in future releases of Openwrt" > "/dev/stderr"
```
I would leave the `exit(1)` but comment it out.  In the future, when users have had a chance to correct their configurations, then we can start enforcing it with the `exit`.

yogo1212 replied:
please resolve if you're happy with the change

pprindeville commented on package/network/ipv6/6rd/files/6rd.sh:
```diff
@@ -9,6 +9,11 @@
 	init_proto "$@"
 }
 
+ipcalc() {
```
Let's put this into `/lib/functions.sh`.

yogo1212 replied:
:+1:
Nice - give me a sec

yogo1212 replied:
Like that?
You can just resolve if there's nothing left to do.

pprindeville commented on package/network/services/dnsmasq/files/dnsmasq.init:
```diff
@@ -509,6 +509,10 @@ dhcp_boot_add() {
 	dhcp_option_add "$cfg" "$networkid" "$force"
 }
 
+ipcalc() {
```
... and that would avoid duplication.

pprindeville reviewed (changes requested):
Minor change: put `ipcalc()` into `/lib/functions.sh` otherwise, looks good.

Ansuel commented:
@pprindeville thanks for taking care of this!

pprindeville commented:
> @pprindeville thanks for taking care of this!

Happy to do it.  It's also on my path to writing the Kea backend for the existing isc-dhcp UCI configs...

yogo1212 commented:
I've squashed "package: dnsmasq: handle return value of ipcalc.sh" into the commit adding the new function because it effectively fixes the bug already (in this new timeline).

yogo1212 commented:
Actual changes (before rebase):
https://github.com/openwrt/openwrt/compare/4105b5cca4e434f9ef4629df4e1d2b6744edc387..b973b2ec07124e7ca10f1ff84198fb64fd56176c

pprindeville reviewed (approved):
Ship it!
